### PR TITLE
Rover hold/loiter

### PR DESCRIPTION
**Pull request by ealdaz-seesai**
This is my first pull request, so unsure as to whether I should  raise an Issue before creating a pull request, but I'm hoping that this explanation will be sufficiently clear, any feedback welcome. 

**Describe problem solved by this pull request**
When using HOLD/LOITER the rover would not hold position but instead circle/oscillate around the hold position. This was caused by the fact that the current algorithm sets velocity to zero when distance to waypoint is under a certain threshold, the vehicle would decelerate, but not enough to stop before it once again exceeded the threshold. At this point it would then try to return to the hold position and the process would repeat. 

**Describe your solution**
In the proposed solution once the distance to the waypoint is under a threshold the velocity is set to zero until a new position waypoint is requested that is more than the threshold distance away from the previous waypoint. This way the vehicle always stops.

**Test data / coverage**
This was tested with both px4_sitl and a pixhawk4 rover running the  Aion RObotics R1 UGV (Tank type/Differential drive)

### CODE REVIEW

**Comment by ealdaz-seesai** [comment]
This is an example of what happens when selecting HOLD in QGroundControl
image

**Comment by ealdaz-seesai** [comment]
In the proposed implementation the rover stops when HOLD is requested:
image

**Comment by julianoes** [comment]
@ItsTimmy would be great if you could review this one, thanks!

File: src/modules/rover_pos_control/RoverPositionControl.cpp [hunk 1]
```diff
@@ -513,12 +518,13 @@ RoverPositionControl::run()
...
-			    manual_mode) {
-				/* publish the actuator controls */
-				_actuator_controls_pub.publish(_act_controls);
+			if (_actuator_controls_pub != nullptr) {
+				//PX4_INFO("Publishing actuator from pos control");
+				orb_publish(ORB_ID(actuator_controls_0), _actuator_controls_pub, &_act_controls);
+
+			} else {
+
+				_actuator_controls_pub = orb_advertise(ORB_ID(actuator_controls_0), &_act_controls);
```
Comment (ItsTimmy) [inline src/modules/rover_pos_control/RoverPositionControl.cpp]: Why was this changed? Shouldn't we still only publish actuator controls when in the right mode?

**Comment by julianoes** [comment]
@ealdaz-seesai it would be great if you could follow up on the things @ItsTimmy mentioned and make sure the CI tests pass.

**Comment by ealdaz-seesai** [comment]
Will do Julian, currently travelling with very bad coverage. Back Monday.
Thanks for pushing this along.

Eduardo

---

Eduardo Aldaz Carroll
CTO
sees.ai

E: ealdaz@sees.ai
M: +44 (0) 7543 519191
Skype: e.aldaz-carroll

File: src/modules/rover_pos_control/RoverPositionControl.hpp [hunk 2]
```diff
@@ -99,10 +98,10 @@ class RoverPositionControl : public ModuleBase<RoverPositionControl>, public Mod
 	void run() override;
 
 private:
-	uORB::Publication<position_controller_status_s>	_pos_ctrl_status_pub{ORB_ID(position_controller_status)};
-	uORB::Publication<actuator_controls_s>		_actuator_controls_pub{ORB_ID(actuator_controls_0)};
+	orb_advert_t	_pos_ctrl_status_pub{nullptr};		/**< navigation capabilities publication */
+	orb_advert_t    _actuator_controls_pub{nullptr};	/**< actuator controls publication */
```
[the remark this answers is not in the document]
Comment (ealdaz-seesai) [reply src/modules/rover_pos_control/RoverPositionControl.hpp]: To be honest I found examples of both styles in the repository, so from memory I used whatever was been used for Rover. Can you modify them to be correct, or is that something I need to do on my side?

File: src/modules/rover_pos_control/RoverPositionControl.cpp (hunk 1 above)
Comment (ealdaz-seesai) [reply src/modules/rover_pos_control/RoverPositionControl.cpp]: As I understand it RoverPositionControl is only called when we are actively in position mode,  see line 422

```
if (!manual_mode && _control_mode.flag_control_position_enabled) {
if (control_position(current_position, ground_speed, _pos_sp_triplet)) {
```

So didn't really understand why the flags for velocity or attitude control are checked inside the position loop, perhaps it's inherited from an older implementation and hasn't been cleaned before?

Comment (Jaeyoung-Lim) [reply src/modules/rover_pos_control/RoverPositionControl.cpp]: @ealdaz-seesai This is not true, although the name of the file is rover position control, it can be used in different modes such as offboard modes. Those functions are used to support different setpoints(position, velocity, attitude, actuator) in offboard mode.

File: src/modules/rover_pos_control/RoverPositionControl.cpp [hunk 3]
```diff
@@ -462,7 +462,12 @@ RoverPositionControl::run()
 
 					pos_ctrl_status.timestamp = hrt_absolute_time();
 
-					_pos_ctrl_status_pub.publish(pos_ctrl_status);
+					if (_pos_ctrl_status_pub != nullptr) {
+						orb_publish(ORB_ID(position_controller_status), _pos_ctrl_status_pub, &pos_ctrl_status);
+
+					} else {
+						_pos_ctrl_status_pub = orb_advertise(ORB_ID(position_controller_status), &pos_ctrl_status);
+					}
```
[the remark this answers is not in the document]
Comment (ealdaz-seesai) [reply src/modules/rover_pos_control/RoverPositionControl.cpp]: Done

File: src/modules/rover_pos_control/RoverPositionControl.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ealdaz-seesai) [reply src/modules/rover_pos_control/RoverPositionControl.hpp]: Changed to C++ API

**Comment by Jaeyoung-Lim** [comment]
@ealdaz-seesai Could you resolve the conflicts of this PR?

**Comment by ealdaz-seesai** [comment]
Of course, hadn't realized there were conflicts, might be due to some new changes to the file. Let me check.

**Comment by ealdaz-seesai** [comment]
All tests passing except some MacOS, complaining about implicit conversion from float to double in calls to function get_distance_to_next_waypoint  using Vector2f variables

**Comment by ealdaz-seesai** [comment]
Is there anything else required to merge this?

**Comment by ealdaz-seesai** [comment]
bump 
@ItsTimmy @dagar @julianoes @Jaeyoung-Lim anything else necessary to merge this?
It seems to me it's ready to go?
Thanks!

**Comment by ealdaz-seesai** [comment]
Hello Julian
I merged 3 weeks ago as some of the files that I had originally modified 4 months ago had been changed since. All tests passed at the time, and couldn't see any conflicts.
Regardless I've merged again without any conflicts.
All checks passed except for the `Jenkins Compile MacOS` one. 
I don't understand what the error is, doesn't seem to be related to my code for what I can tell.
Could you check please?
Thx

**Comment by ealdaz-seesai** [comment]
Also not sure I understand the submodule point you were making, I haven't changed any submodule.

**Comment by julianoes** [comment]
I'm really confused about the state of this PR, I can't get it merged properly either. It's odd.

Some submodules are wrong, see here:
Selection_048

Let me try to apply the diff and create one new clean commit.

**Comment by julianoes** [comment]
I have the diff applied cleanly but I can't push it to your fork.

Could you make the checkmark as described here?
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by ealdaz-seesai** [comment]
I’m really sorry Julian, but I can’t find that check box anywhere on the Pull request page….

Any tips on how to find it, it’s like it’s not been enabled ...




Eduardo

---

Eduardo Aldaz Carroll
CTO
sees.ai

E: ealdaz@sees.ai
M: +44 (0) 7543 519191
Skype: e.aldaz-carroll

**Comment by julianoes** [comment]
> Is it because the repository is not my personal one but part of an organisation?

That's probably why. Do you want to create a new PR with the clean diff or should I do it?

**Comment by ealdaz-seesai** [comment]
I’ll give it a go

Eduardo

---

Eduardo Aldaz Carroll
CTO
sees.ai

E: ealdaz@sees.ai
M: +44 (0) 7543 519191
Skype: e.aldaz-carroll

**Comment by ealdaz-seesai** [comment]
OK, I've figured out what the issue was. Those submodules were added to my git branch 3 weeks ago and I didn't realize. I've removed them now. I'm hoping it should all be OK now.

**Comment by julianoes** [comment]
Great, thanks @ealdaz-seesai.

@Jaeyoung-Lim could you please give this a review?

File: src/modules/rover_pos_control/RoverPositionControl.hpp [hunk 4]
```diff
@@ -113,15 +114,15 @@ class RoverPositionControl : public ModuleBase<RoverPositionControl>, public Mod
...
 	position_setpoint_triplet_s		_pos_sp_triplet{};		/**< triplet of mission items */
 	vehicle_attitude_setpoint_s		_att_sp{};			/**< attitude setpoint > */
 	vehicle_control_mode_s			_control_mode{};		/**< control mode */
 	vehicle_global_position_s		_global_pos{};			/**< global vehicle position */
 	vehicle_local_position_s		_local_pos{};			/**< global vehicle position */
-	actuator_controls_s			_act_controls{};		/**< direct control of actuators */
-	vehicle_attitude_s			_vehicle_att{};
-	sensor_combined_s			_sensor_combined{};
+	actuator_controls_s			    _act_controls{};		/**< direct control of actuators */
+	vehicle_attitude_s              _vehicle_att{};
```
Comment (Jaeyoung-Lim) [inline src/modules/rover_pos_control/RoverPositionControl.hpp:124]: Please fix the indentation errors

File: src/modules/rover_pos_control/RoverPositionControl.hpp [hunk 5]
```diff
@@ -113,15 +114,15 @@ class RoverPositionControl : public ModuleBase<RoverPositionControl>, public Mod
 
 	uORB::Subscription	_parameter_update_sub{ORB_ID(parameter_update)};
 
-	manual_control_setpoint_s		_manual{};			/**< r/c channel data */
+	manual_control_setpoint_s		_manual{};			    /**< r/c channel data */
```
Comment (Jaeyoung-Lim) [inline src/modules/rover_pos_control/RoverPositionControl.hpp:117]: unnecessary indentation changes

File: src/modules/rover_pos_control/RoverPositionControl.hpp [hunk 6]
```diff
@@ -113,15 +114,15 @@ class RoverPositionControl : public ModuleBase<RoverPositionControl>, public Mod
...
 	vehicle_attitude_setpoint_s		_att_sp{};			/**< attitude setpoint > */
 	vehicle_control_mode_s			_control_mode{};		/**< control mode */
 	vehicle_global_position_s		_global_pos{};			/**< global vehicle position */
 	vehicle_local_position_s		_local_pos{};			/**< global vehicle position */
-	actuator_controls_s			_act_controls{};		/**< direct control of actuators */
-	vehicle_attitude_s			_vehicle_att{};
-	sensor_combined_s			_sensor_combined{};
+	actuator_controls_s			    _act_controls{};		/**< direct control of actuators */
+	vehicle_attitude_s              _vehicle_att{};
+	sensor_combined_s				_sensor_combined{};
```
Comment (Jaeyoung-Lim) [inline src/modules/rover_pos_control/RoverPositionControl.hpp:125]: same here

File: src/modules/rover_pos_control/RoverPositionControl.cpp [hunk 7]
```diff
@@ -406,7 +407,7 @@ RoverPositionControl::run()
 		bool manual_mode = _control_mode.flag_control_manual_enabled;
 
 		/* only run controller if position changed */
-		if (fds[0].revents & POLLIN) {
+		if (fds[0].revents & POLLIN || fds[4].revents & POLLIN) {
```
Comment (Jaeyoung-Lim) [inline src/modules/rover_pos_control/RoverPositionControl.cpp:410]: I think we can use just local_position (`fds[4]`) and remove `fds[0]`

File: src/modules/rover_pos_control/RoverPositionControl.hpp (hunk 4 above)
Comment (ealdaz-seesai) [reply src/modules/rover_pos_control/RoverPositionControl.hpp:124]: OK, will do. 
I run the fix_code_style.sh, thought that took care of that sort of thing. But just realized that CLion was using tabs as spaces, will sort it out.

File: src/modules/rover_pos_control/RoverPositionControl.cpp (hunk 7 above)
Comment (ealdaz-seesai) [reply src/modules/rover_pos_control/RoverPositionControl.cpp:410]: Ha, pushing my very average git skills with this...never tried squashign commits before, I'll google and give it a try...

Comment (ealdaz-seesai) [reply src/modules/rover_pos_control/RoverPositionControl.cpp:410]: Is local position also available if only using GPS?

**Comment by ealdaz-seesai** [comment]
Hi @Jaeyoung-Lim , Ok so I've really tried, I've rebased and squashed, but still all the commits appear. I tried rebasing again and then all sorts of conflicts happen, I think it's got to do probably with the fact that I've merged twice already in the last months. I'm not comfortable enough with git to understand what is really going on, and I'm afraid I can't really spend anymore time with this.

**Comment by ealdaz-seesai** [comment]
OK, so it was annoying me to leave it half done. So decided to give it one last go and rebase again and skip some commits and it has worked.
I've tested in SITL and all works as intended.

I did notice a new "feature" in this latest master and it's that SITL rover complains that no power is unavailable and doesn't let you arm. I had to disable power check flag. (CBRK_SUPPLY_CHECK).

Note that I couldn't reboot either from QGroundcontrol... i'm pretty sure I could before... 
So a couple of things to add as a future PR...
```
pxh> WARN  [PreFlightCheck] system power unavailable
````

File: src/modules/rover_pos_control/RoverPositionControl.hpp (hunk 4 above)
Comment (Jaeyoung-Lim) [reply src/modules/rover_pos_control/RoverPositionControl.hpp]: Thanks!

**Comment by ealdaz-seesai** [comment]
Jenkins SITL Tests fail, but not sure why... would any of you know ?

**Comment by julianoes** [comment]
@Jaeyoung-Lim would be great if you could follow up.

File: src/modules/rover_pos_control/RoverPositionControl.hpp (hunk 5 above)
Comment (Jaeyoung-Lim) [reply src/modules/rover_pos_control/RoverPositionControl.hpp:117]: White spaces are still not addressed

File: src/modules/rover_pos_control/RoverPositionControl.hpp (hunk 6 above)
Comment (Jaeyoung-Lim) [reply src/modules/rover_pos_control/RoverPositionControl.hpp:125]: White spaces are still not addressed

**Comment by Jaeyoung-Lim** [comment]
@julianoes Jenkins failure is not related to this PR. There are a few comments that are not addressed from @ealdaz-seesai Otherwise looks good to me

**Comment by julianoes** [comment]
@Jaeyoung-Lim you're call if you want those addressed or merge anyway.

**Comment by julianoes** [comment]
Thanks @ealdaz-seesai !

**Comment by ealdaz-seesai** [comment]
Thank you both !  A pleasure!

Eduardo

---

Eduardo Aldaz Carroll
CTO
sees.ai

E: ealdaz@sees.ai
M: +44 (0) 7543 519191
Skype: e.aldaz-carroll
